nrf_rpc/include/nrf_rpc.h

```diff
@@ -0,0 +1,444 @@
+ *                  encoded data.
+ * @param len       Length of the packet. Can be smaller than allocated.
+ * @returns         NRF_RPC_SUCCESS or error code from enum nrf_rpc_error_code.
+ */
+int nrf_rpc_evt_send(nrf_rpc_alloc_ctx ctx, uint8_t evt, uint8_t *packet,
+		     size_t len);
+
+
+/** @brief Send an event passing any errors to an error handler.
+ *
+ * If error occurred during sending this event it will be passed to
+ * @a nrf_rpc_error_handler instead of returning it. This form of error handling
+ * can be useful for serializing API function that has no ability to report
+ * an error in any other way, e.g. it returns void.
+ *
+ * See @a nrf_rpc_evt_send for more details.
+ */
+void nrf_rpc_evt_send_noerr(nrf_rpc_alloc_ctx ctx, uint8_t evt, uint8_t *packet,
+			    size_t len);
+
+
+/** @brief Send a response to a command.
+ *
+ * There is no _noerr form of this function, because it is always called from
+ * command decoder, so the error code returned by @a nrf_rpc_rsp_send can
+ * be passed further and returned from command decoder. Error returned from
+ * commands decoders always go to @a nrf_rpc_error_handler.
+ *
+ * @param remote_ep Destination endpoint allocated by @a NRF_RPC_RSP_ALLOC.
+ * @param packet    Packet allocated by @a NRF_RPC_RSP_ALLOC and filled with
+ *                  encoded data.
+ * @param len       Length of the packet. Can be smaller than allocated.
+ * @returns         NRF_RPC_SUCCESS or error code from enum nrf_rpc_error_code.
+ */
+int nrf_rpc_rsp_send(nrf_rpc_alloc_ctx ctx, uint8_t *packet, size_t len);
+
+
+/** @brief Indicate that decoding of the input data is done.
+ *
+ * This function must be called from a command decoder and an event decoder
+ * as soon as the input was parsed. If a response is decoded inside a command
+ * encoder (i.e. using @a nrf_rpc_cmd_rsp_send) then this function also must
+ * be called. If a response is decoded in separete handler (i.e. using
+ * @a nrf_rpc_cmd_send or @a nrf_rpc_cmd_send_noerr) then this function cannot
+ * be called. Content of the input data is no longer valid after this function
+ * call.
+ */
+void nrf_rpc_decoding_done(void);
+
+
+/** @brief Function for handling errors that cannot be exposed as an API return
+ * value.
+ *
+ * Some errors (e.g. corrupted packet received) cannot be easily returned by a
+ * function, because there was no actual function call from user.
+ * `nrf_rpc_error_handler` function will be called in case of such error. It is
+ * weakly defined as empty function, so user can provide a different
+ * implementation.
+ *
+ * @param tr_local_ep  Local endpoint associated with an error or NULL if it is
+ *                     unknown.
+ * @param tr_remote_ep Remote encpoint associated with an error or NULL if it is
+ *                     unknown.
+ * @param code         Error code
+ * @param from_remote  If error occurred on the remote side.
+ */
+void nrf_rpc_error_handler(struct nrf_rpc_tr_local_ep *tr_local_ep,
```

Comment:
Is it possible to figure out which command caused this error?

Reply:
Not always, e.g. when we received malformed packet, but in some cases yes. Do you think that adding another parameters with command and group id will be beneficial?

Reply:
Yes, sure, I think it would be easier to at least identify command that failed. If there are too many parameters in this handler, we can use a structure to group error info.

Reply:
I put it into a structure `struct nrf_rpc_err_report` as you suggested. I also added `NRF_RPC_ID_UNKNOWN` definition to mark that something (e.g. cmd id) is unknown.